src/features/codeLensProvider.ts

```diff
@@ -6,171 +6,276 @@
 import * as protocol from '../omnisharp/protocol';
 import * as serverUtils from '../omnisharp/utils';
 import * as vscode from 'vscode';
-import { toLocation, toRange } from '../omnisharp/typeConvertion';
+import { toLocation } from '../omnisharp/typeConvertion';
 import AbstractProvider from './abstractProvider';
 import { OmniSharpServer } from '../omnisharp/server';
 import { Options } from '../omnisharp/options';
 import TestManager from './dotnetTest';
 import OptionProvider from '../observers/OptionProvider';
 
-class OmniSharpCodeLens extends vscode.CodeLens {
+import Structure = protocol.V2.Structure;
+import SymbolKinds = protocol.V2.SymbolKinds;
+import SymbolPropertyNames = protocol.V2.SymbolPropertyNames;
+import SymbolRangeNames = protocol.V2.SymbolRangeNames;
 
-    fileName: string;
+abstract class OmniSharpCodeLens extends vscode.CodeLens {
+    constructor(
+        range: protocol.V2.Range,
+        public fileName: string) {
 
-    constructor(fileName: string, range: vscode.Range) {
-        super(range);
-        this.fileName = fileName;
+        super(new vscode.Range(
+            range.Start.Line - 1, range.Start.Column - 1, range.End.Line - 1, range.End.Column - 1
+        ));
+    }
+}
+
+class ReferencesCodeLens extends OmniSharpCodeLens {
+    constructor(
+        range: protocol.V2.Range,
+        fileName: string) {
+            super(range, fileName);
+    }
+}
+
+abstract class TestCodeLens extends OmniSharpCodeLens {
+    constructor(
+        range: protocol.V2.Range,
+        fileName: string, 
+        public testFramework: string,
+        public testMethodNames: string[]) {
+
+        super(range, fileName);
+    }
+}
+
+class RunTestsCodeLens extends TestCodeLens {
+    constructor(
+        range: protocol.V2.Range,
+        fileName: string, 
+        testFramework: string,
+        testMethodNames: string[]) {
+
+        super(range, fileName, testFramework, testMethodNames);
+    }
+}
+
+class DebugTestsCodeLens extends TestCodeLens {
+    constructor(
+        range: protocol.V2.Range,
+        fileName: string, 
+        testFramework: string,
+        testMethodNames: string[]) {
+
+        super(range, fileName, testFramework, testMethodNames);
     }
 }
 
 export default class OmniSharpCodeLensProvider extends AbstractProvider implements vscode.CodeLensProvider {
 
     constructor(server: OmniSharpServer, testManager: TestManager, private optionProvider: OptionProvider) {
         super(server);
-
     }
 
-    private static filteredSymbolNames: { [name: string]: boolean } = {
-        'Equals': true,
-        'Finalize': true,
-        'GetHashCode': true,
-        'ToString': true
-    };
-
-    async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken) {
-        let options = this.optionProvider.GetLatestOptions();
+    async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken): Promise<vscode.CodeLens[]> {
```

Comment:
I'm sure there are zero tests for this, right 😄

Reply:
yes, I was planning to add some.

Reply:
Added integration tests for both DocumentSymbolProvider and CodeLensProvider